apps/controller/main.cpp

```diff
@@ -0,0 +1,132 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include "AnalogInputGroup.h"
+#include "DigitalInputGroup.h"
+#include "FrameController.h"
+#include "mbed.h"
+
+using namespace FrameProtocol;
+
+#define DEBUG
+
+static constexpr uint16_t SEND_INTERVAL_MS         = 100;
+static constexpr uint8_t DEBOUNCE_THRES            = 3;
+static constexpr uint8_t ANALOG_NUM_SAMPLE_AVERAGE = 5;
+
+static constexpr uint8_t NUM_BTNS     = 3;
+static constexpr uint8_t NUM_SWITCHES = 4;
+static constexpr uint8_t NUM_JOYS     = 4;
+static constexpr uint8_t NUM_POTS     = 2;
+
+BusIn digital_inputs_bus_btns(JOY_BTN, PB1, PB2);
+BusIn digital_inputs_bus_switches(SW_1A, SW_1B, SW_2A, SW_2B);
+
+// we can tune these parameters while testing
+DigitalInputGroup btns(digital_inputs_bus_btns, NUM_BTNS, InputDebounceType::INTERGRATING, DEBOUNCE_THRES);
+DigitalInputGroup switches(digital_inputs_bus_switches, NUM_SWITCHES, InputDebounceType::INTERGRATING, DEBOUNCE_THRES);
+
+FrameController::DigitalFrameConfig btns_frame_config = {
+    .boardType  = BoardType::MASTER,
+    .inputType  = DigitalInputType::BTN,
+    .inputGroup = btns,
+};
+
+FrameController::DigitalFrameConfig switches_frame_config = {
```

Comment:
This might be a bit confusing ... so the namespace is named as FrameController and the class is also named as FrameController, the DigitalFrameConfig is declared inside the FrameController class....maybe I should rename the namespace to sth else

Reply:
I agree lol